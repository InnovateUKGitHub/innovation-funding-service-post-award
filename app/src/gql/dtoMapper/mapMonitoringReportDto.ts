import {
  MonitoringReportDto,
  MonitoringReportQuestionGqlDto,
  MonitoringReportQuestionDto,
} from "@framework/dtos/monitoringReportDto";
import { groupBy } from "lodash";
import { mapMonitoringReportStatus } from "@framework/util/monitoringReportStatus";

type MonitoringReportNode = Readonly<
  Partial<{
    Id: string;
    Acc_AddComments__c: GQL.Value<string>;
    Acc_FinalMonitoringReport__c: GQL.Value<boolean>;
    Acc_MonitoringHeader__c: GQL.Value<string>;
    Acc_MonitoringReportStatus__c: {
      value: string | null;
      label: string | null;
    } | null;
    Acc_PeriodEndDate__c: GQL.Value<string>;
    Acc_PeriodStartDate__c: GQL.Value<string>;
    Acc_Project__c: GQL.Value<string>;
    Acc_ProjectPeriodNumber__c: GQL.Value<number>;
    Acc_Question__c: GQL.Value<string>;
    Acc_QuestionComments__c: GQL.Value<string>;
    LastModifiedDate: GQL.Value<string>;
    RecordType: {
      Name: GQL.Value<string>;
    } | null;
  }>
> | null;

type MonitoringReportDtoMapping = Pick<
  MonitoringReportDto,
  | "addComments"
  | "endDate"
  | "headerId"
  | "lastUpdated"
  | "periodId"
  | "projectId"
  | "startDate"
  | "status"
  | "statusName"
  | "questions"
>;

const HEADER = "Monitoring Header";
const ANSWER = "Monitoring Answer";

const updatableStatuses = ["New", "Draft", "IUK Queried"];

const headerMapper: GQL.DtoMapper<Omit<MonitoringReportDtoMapping, "questions">, MonitoringReportNode> = {
  addComments(node) {
    return node?.Acc_AddComments__c?.value ?? null;
  },
  status(node) {
    return mapMonitoringReportStatus(node?.Acc_MonitoringReportStatus__c?.value ?? "unknown");
  },
  headerId(node) {
    return (
      node?.RecordType?.Name?.value === ANSWER
        ? node?.Acc_MonitoringHeader__c?.value ?? "unknown-monitoring-report-id"
        : node?.Id ?? ""
    ) as MonitoringReportId;
  },
  periodId(node) {
    return (node?.Acc_ProjectPeriodNumber__c?.value ?? 0) as PeriodId;
  },
  startDate(node) {
    return !!node?.Acc_PeriodStartDate__c?.value ? new Date(node?.Acc_PeriodStartDate__c?.value) : null;
  },
  endDate(node) {
    return !!node?.Acc_PeriodEndDate__c?.value ? new Date(node?.Acc_PeriodEndDate__c?.value) : null;
  },
  statusName(node) {
    return node?.Acc_MonitoringReportStatus__c?.label ?? "unknown-status";
  },
  lastUpdated(node) {
    return !!node?.LastModifiedDate?.value ? new Date(node?.LastModifiedDate?.value) : null;
  },
  projectId(node) {
    return (node?.Acc_Project__c?.value ?? "") as ProjectId;
  },
};

/**
 * Maps a specified Monitoring Report Node from a GQL query to a slice of
 * the MonitoringReportDto to ensure consistency and compatibility in the application
 */
export function mapToMonitoringReportDto<
  T extends MonitoringReportNode,
  TPickList extends keyof MonitoringReportDtoMapping,
>(node: T, pickList: TPickList[]): Pick<MonitoringReportDtoMapping, TPickList> {
  return pickList.reduce((dto, field) => {
    if (field === "questions") return dto;

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore don't know how to let it know that "questions" will never be applied here
    dto[field] = headerMapper[field as Exclude<TPickList, "questions">](node);
    return dto;
  }, {} as Pick<MonitoringReportDtoMapping, TPickList>);
}

/**
 * Maps edges to array of MonitoringReport DTOs.
 */
export function mapToMonitoringReportDtoArray<
  T extends ReadonlyArray<{ node: MonitoringReportNode } | null> | null,
  TPickList extends keyof MonitoringReportDtoMapping,
>(edges: T, pickList: TPickList[]): Pick<MonitoringReportDtoMapping, TPickList>[] {
  return (
    edges?.map(x => {
      return mapToMonitoringReportDto(x?.node ?? null, pickList);
    }) ?? []
  );
}

/**
 * gets the Active questions for monitoring reports that can be edited, or
 * the Answered questions for readonly monitoring report
 */
function getQuestions(
  header: Pick<MonitoringReportDto, "statusName">,
  results: ({ node: MonitoringReportNode | null } | null)[],
  questions: Pick<
    MonitoringReportQuestionGqlDto,
    "isActive" | "title" | "displayOrder" | "id" | "isScored" | "description" | "options"
  >[],
) {
  if (updatableStatuses.includes(header.statusName)) {
    // map Active Questions
    const mappedQuestions = groupBy(
      questions.filter(x => x.isActive).sort((a, b) => a.displayOrder - b.displayOrder),
      "displayOrder",
    );

    const activeQuestions = Object.values(mappedQuestions).map(options => ({
      title: options[0].title,
      displayOrder: options[0].displayOrder,
      optionId: !options[0].isScored ? options[0].id : null,
      responseId: null,
      comments: null,
      description: options[0].description,
      isScored: options[0].isScored,
      options: options
        .map(o => {
          return {
            id: o.options[0].id,
            questionText: o.options[0].questionText,
            questionScore: o.options[0].questionScore,
          };
        })
        // The options should be displayed in descending score order (largest at the top of the list of options)
        .sort((a, b) => b.questionScore - a.questionScore),
    }));

    return activeQuestions;
  }

  // Gets the answered questions
  const questionIds = results.map(x => x?.node?.Acc_Question__c?.value);

  return questions
    .filter(x => questionIds.includes(x.id))
    .sort((a, b) => a.displayOrder - b.displayOrder)
    .map(x => ({
      ...x,
      optionId: null,
      responseId: null,
      comments: null,
    }));
}

/**
 * maps comments and ids onto the monitoring report questions
 */
function populateAnswer<T extends Pick<MonitoringReportQuestionDto, "options">>(
  question: T,
  results: ({ node: MonitoringReportNode | null } | null)[],
): T {
  const options = question.options.map(o => o.id);
  // if there are no options get it from the preselected answer as its a non-option question
  const response = (results ?? []).find(r => options.includes(r?.node?.Acc_Question__c?.value ?? "unknown"));

  if (!response) {
    return question;
  }

  return Object.assign(question, {
    optionId: response?.node?.Acc_Question__c?.value ?? null,
    comments: response?.node?.Acc_QuestionComments__c?.value ?? null,
    responseId: response?.node?.Id,
  });
}

/**
 * maps monitoring report data to the header and linked questions
 */
export function mapToFullMonitoringReport<
  T extends ReadonlyArray<{ node: MonitoringReportNode } | null> | null,
  TPickList extends keyof MonitoringReportDtoMapping,
>(
  edges: T,
  pickList: "statusName" extends TPickList ? TPickList[] : never, // It is required to pass in the "statusName" to the picklist
  additionalData: {
    questions: Pick<
      MonitoringReportQuestionGqlDto,
      "isActive" | "title" | "displayOrder" | "id" | "isScored" | "description" | "options"
    >[];
  },
): Pick<MonitoringReportDtoMapping, TPickList> {
  if (!pickList.includes("statusName" as TPickList)) {
    throw new Error("statusName is a required field for fetching the full monitoring report");
  }
  const headerGql = edges && edges.length > 1 && edges.find(x => x?.node?.RecordType?.Name?.value === HEADER);
  if (!headerGql) throw new Error("Invalid Monitoring Report Request - no matching Header");
  const header = mapToMonitoringReportDto(headerGql.node, pickList);

  const results = (edges ?? []).filter(x => x?.node?.RecordType?.Name?.value === ANSWER);

  const questions = getQuestions(
    header as Pick<MonitoringReportDto, "statusName">,
    results,
    additionalData.questions,
  ).map(x => populateAnswer(x, results));

  return {
    ...header,
    questions,
  };
}
