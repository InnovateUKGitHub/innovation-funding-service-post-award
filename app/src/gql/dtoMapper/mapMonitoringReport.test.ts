import { mapToMonitoringReportDtoArray } from "./mapMonitoringReportDto";

describe("mapToMonitoringReportDtoArray", () => {
  const edges = [
    {
      node: {
        Id: "0",
        Acc_MonitoringReportStatus__c: {
          value: "Draft",
          label: "Draft label",
        },
        Acc_PeriodStartDate__c: { value: "2022-02-24" },
        Acc_PeriodEndDate__c: { value: "2022-04-24" },
        Acc_FinalMonitoringReport__c: { value: true },
        Acc_ProjectPeriodNumber__c: { value: 1 },
        LastModifiedDate: { value: "2022-04-24" },
        Acc_Project__c: { value: "1" },
      },
    },
    {
      node: {
        Id: "1",
        Acc_MonitoringReportStatus__c: {
          value: "Approved",
          label: "Approved label",
        },
        Acc_PeriodStartDate__c: { value: "2022-02-24" },
        Acc_PeriodEndDate__c: { value: "2022-04-24" },
        Acc_FinalMonitoringReport__c: { value: false },
        Acc_ProjectPeriodNumber__c: { value: 2 },
        LastModifiedDate: { value: "2022-04-24" },
        Acc_Project__c: { value: "1" },
      },
    },
    {
      node: {
        Id: "2",
        Acc_MonitoringReportStatus__c: {
          value: "Awaiting IUK Approval",
          label: "Awaiting IUK Approval label",
        },
        Acc_PeriodStartDate__c: { value: "2022-02-24" },
        Acc_PeriodEndDate__c: { value: "2022-04-24" },
        Acc_FinalMonitoringReport__c: { value: true },
        Acc_ProjectPeriodNumber__c: { value: 3 },
        LastModifiedDate: { value: "2022-04-24" },
        Acc_Project__c: { value: "1" },
      },
    },
    {
      node: {
        Id: "3",
        Acc_MonitoringReportStatus__c: null,
        Acc_PeriodStartDate__c: null,
        Acc_PeriodEndDate__c: null,
        Acc_FinalMonitoringReport__c: null,
        Acc_ProjectPeriodNumber__c: null,
        LastModifiedDate: null,
        Acc_Project__c: null,
      },
    },
  ];

  it("should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records", () => {
    expect(
      mapToMonitoringReportDtoArray(edges, [
        "headerId",
        "status",
        "periodId",
        "startDate",
        "endDate",
        "statusName",
        "lastUpdated",
        "projectId",
      ]),
    ).toMatchSnapshot();
  });

  it("should map the gql data only including fields on the picklist", () => {
    expect(mapToMonitoringReportDtoArray(edges, ["headerId", "status", "periodId"])).toMatchSnapshot();
  });
});
