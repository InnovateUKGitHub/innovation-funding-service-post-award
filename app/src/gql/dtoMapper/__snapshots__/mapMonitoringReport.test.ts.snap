// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapToMonitoringReportDtoArray should map the gql data only including fields on the picklist 1`] = `
Array [
  Object {
    "headerId": "0",
    "periodId": 1,
    "status": 20,
  },
  Object {
    "headerId": "1",
    "periodId": 2,
    "status": 40,
  },
  Object {
    "headerId": "2",
    "periodId": 3,
    "status": 30,
  },
  Object {
    "headerId": "3",
    "periodId": 0,
    "status": 0,
  },
]
`;

exports[`mapToMonitoringReportDtoArray should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records 1`] = `
Array [
  Object {
    "endDate": 2022-04-24T00:00:00.000Z,
    "headerId": "0",
    "lastUpdated": 2022-04-24T00:00:00.000Z,
    "periodId": 1,
    "projectId": "1",
    "startDate": 2022-02-24T00:00:00.000Z,
    "status": 20,
    "statusName": "Draft label",
  },
  Object {
    "endDate": 2022-04-24T00:00:00.000Z,
    "headerId": "1",
    "lastUpdated": 2022-04-24T00:00:00.000Z,
    "periodId": 2,
    "projectId": "1",
    "startDate": 2022-02-24T00:00:00.000Z,
    "status": 40,
    "statusName": "Approved label",
  },
  Object {
    "endDate": 2022-04-24T00:00:00.000Z,
    "headerId": "2",
    "lastUpdated": 2022-04-24T00:00:00.000Z,
    "periodId": 3,
    "projectId": "1",
    "startDate": 2022-02-24T00:00:00.000Z,
    "status": 30,
    "statusName": "Awaiting IUK Approval label",
  },
  Object {
    "endDate": null,
    "headerId": "3",
    "lastUpdated": null,
    "periodId": 0,
    "projectId": "",
    "startDate": null,
    "status": 0,
    "statusName": "unknown",
  },
]
`;
